ki/src/Ki/Internal/Scope.hs

```diff
@@ -90,7 +91,8 @@ data Scope = Scope
     -- can continue to delay; there's no opportunity for an async exception to strike and prevent one of these threads
     -- from starting.
     --
-    -- Sentinel value: -1 means the scope is closed.
+    -- Sentinel value: -1 means the scope is closing.
+    -- Sentinel value: -2 means the scope is closed.
```

Comment:
Why do you use sentinel values instead of a proper sum type?

Reply:
Hm, I think it'd be good to verify that! One sentinel value is liveable, but two starts getting very hairy! But then again, it's not my library :laughing:

Reply:
Hehe, good point. Well, I'd like to think of it not as "my" library either! I'll add benching that to the todo list.

Reply:
It would be interesting to know what is the overhead of using a sum type here. Out of curiosity I wrote this playground https://play.haskell.org/saved/r85araK7 , and it looks like doing `n >= 0` produces more core than a pattern match. Though I can't tell if that's better or not :)